File: dynamic_programming/knapsack.js
/**
 * 0/1 Knapsack Problem using Dynamic Programming
 */

function knapsack(W, wt, val) {
    const n = wt.length;
    // Initialize DP table
    const dp = Array(n + 1).fill().map(() => Array(W + 1).fill(0));

    // Build table
    for (let i = 1; i <= n; i++) {
        for (let w = 1; w <= W; w++) {
            if (wt[i - 1] <= w) {
                dp[i][w] = Math.max(
                    val[i - 1] + dp[i - 1][w - wt[i - 1]],
                    dp[i - 1][w]
                );
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    // Reconstruct solution
    const selectedItems = new Set();
    let tempW = W;
    for (let i = n; i > 0; i--) {
        if (dp[i][tempW] !== dp[i - 1][tempW]) {
            selectedItems.add(i);
            tempW -= wt[i - 1];
        }
    }

    return {
        maxValue: dp[n][W],
        selectedItems: selectedItems
    };
}

const val = [3, 2, 4, 4];
const wt = [4, 3, 2, 3];
const W = 6;

const result = knapsack(W, wt, val);
console.log("Optimal Value:", result.maxValue);
console.log("Selected Items:", Array.from(result.selectedItems));

File: dynamic_programming/knapsack.go
package main

import (
	"fmt"
	"math"
)

type Solution struct {
	MaxValue      int
	SelectedItems []int
}

func Knapsack(W int, wt []int, val []int) Solution {
	n := len(wt)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, W+1)
	}

	for i := 1; i <= n; i++ {
		for w := 1; w <= W; w++ {
			if wt[i-1] <= w {
				take := val[i-1] + dp[i-1][w-wt[i-1]]
				skip := dp[i-1][w]
				dp[i][w] = int(math.Max(float64(take), float64(skip)))
			} else {
				dp[i][w] = dp[i-1][w]
			}
		}
	}

	selectedItems := []int{}
	tempW := W
	for i := n; i > 0; i-- {
		if dp[i][tempW] != dp[i-1][tempW] {
			selectedItems = append(selectedItems, i)
			tempW -= wt[i-1]
		}
	}

	return Solution{MaxValue: dp[n][W], SelectedItems: selectedItems}
}

func main() {
	val := []int{3, 2, 4, 4}
	wt := []int{4, 3, 2, 3}
	W := 6

	sol := Knapsack(W, wt, val)
	fmt.Println("Optimal Value:", sol.MaxValue)
	fmt.Println("Selected Items:", sol.SelectedItems)
}
